### PR TITLE
Pyo3 update

### PR DESCRIPTION
**Pull request by awestlake87**
Hey, the fix was a bit more complicated than I anticipated. Essentially, since you're using `run_forever`, there was no way of getting the event loop reference via `get_running_loop` as opposed to `get_event_loop` (which will fetch the global event loop). I went ahead and scoped the event loop for your `index` handler so that it has access to the event loop via task-local data.

I've been improving the docs on 0.14, but this problem probably isn't explicitly addressed. The "Event Loop References and Thread-Awareness" section in the docs may provide some intuition for why 0.14 acts this way.

### CODE REVIEW

File: src/server.rs [hunk 1]
```diff
@@ -44,16 +42,29 @@ impl Server {
         let router = self.router.clone();
         let headers = self.headers.clone();
 
+        let asyncio = py.import("asyncio").unwrap();
+
+        let event_loop = asyncio.call_method0("new_event_loop").unwrap();
```
Comment (ShadowJonathan) [inline src/server.rs:47]: are we sure this is on a python thread? (that will keep executing the event loop with the control on python's side, before we start throwing tasks into it)

Comment (ShadowJonathan) [reply src/server.rs:47]: Oh nevermind, i just saw the `run_forever` pass to python and then asyncio down below, yeah, that should make this alright.

File: src/server.rs [hunk 2]
```diff
@@ -44,16 +42,29 @@ impl Server {
...
+            .unwrap();
+        let event_loop_hdl = PyObject::from(event_loop);
+
         thread::spawn(move || {
             //init_current_thread_once();
             actix_web::rt::System::new().block_on(async move {
                 let addr = format!("127.0.0.1:{}", port);
 
                 HttpServer::new(move || {
+                    let event_loop_hdl = event_loop_hdl.clone();
```
Comment (ShadowJonathan) [inline src/server.rs:59]: Why not just move the value? (and not clone it)

Comment (awestlake87) [reply src/server.rs:59]: `HttpServer::new` expects a `Fn` not a `FnOnce`, so when `event_loop_hdl` gets moved into the router closure later on, `event_loop_hdl` needs to remain valid for the next time the `HttpServer::new` closure is called.

Comment (ShadowJonathan) [reply src/server.rs:59]: Ah, didn't know that, then this LGTM 👍

**Comment by sansyrox** [comment]
Thank you @awestlake87 for this. Do the changes even work with `v0.13.*` ?

**Comment by awestlake87** [comment]
The 0.14 API is pretty different from 0.13, so it won't work with both. If you need it to work with both 0.13 and 0.14, you can delay this PR and continue using the 0.13 API with `#[allow(deprecated)]` since the 0.13 conversions work the same in 0.13 and 0.14.

**Comment by sansyrox** [comment]
@awestlake87 , no no. I was asking as the Cargo.toml was not updated. I will update it and check if the tests pass.

**Comment by sansyrox** [comment]
@awestlake87 , what does `pyo3-asyncio::tokio::scope` do?

**Comment by awestlake87** [comment]
It allows the conversions to get a reference to the current event loop via task-local data. I've got a breakdown of its motivation and usage [here](https://docs.rs/pyo3-asyncio/0.14.0/pyo3_asyncio/#event-loop-references)

**Comment by ShadowJonathan** [comment]
More specifically;

> `pyo3_asyncio::<runtime>::scope` - Store the event loop in task-local data when executing the given Future.

> `pyo3_asyncio::<runtime>::scope_local` - Store the event loop in task-local data when executing the given !Send Future.

**Pull request merged**